src/main/java/org/apache/commons/crypto/stream/CipherInputStream.java

```diff
@@ -70,41 +75,49 @@
   protected ByteBuffer outBuffer;
 
   public CipherInputStream(CipherTransformation transformation,
-                           Properties props, InputStream in, byte[] key, byte[] iv)
+                           Properties props, InputStream in, Key key, AlgorithmParameterSpec params)
       throws IOException {
     this(in, Utils.getCipherInstance(transformation, props),
-        Utils.getBufferSize(props), key, iv);
+        Utils.getBufferSize(props), key, params);
   }
 
   public CipherInputStream(CipherTransformation transformation,
-                           Properties props, ReadableByteChannel in, byte[] key, byte[] iv)
+                           Properties props, ReadableByteChannel in, Key key, AlgorithmParameterSpec params)
       throws IOException {
     this(in, Utils.getCipherInstance(transformation, props),
-        Utils.getBufferSize(props), key, iv);
+        Utils.getBufferSize(props), key, params);
   }
 
   public CipherInputStream(InputStream in, Cipher cipher, int bufferSize,
-                           byte[] key, byte[] iv) throws IOException {
-    this(new StreamInput(in, bufferSize), cipher, bufferSize, key, iv);
+                           Key key, AlgorithmParameterSpec params) throws IOException {
+    this(new StreamInput(in, bufferSize), cipher, bufferSize, key, params);
   }
 
   public CipherInputStream(ReadableByteChannel in, Cipher cipher,
-                           int bufferSize, byte[] key, byte[] iv) throws IOException {
-    this(new ChannelInput(in), cipher, bufferSize, key, iv);
+                           int bufferSize, Key key, AlgorithmParameterSpec params) throws IOException {
+    this(new ChannelInput(in), cipher, bufferSize, key, params);
   }
 
   public CipherInputStream(
-      Input input,
-      Cipher cipher,
-      int bufferSize,
-      byte[] key,
-      byte[] iv) throws IOException {
+          Input input,
+          Cipher cipher,
+          int bufferSize,
+          Key key,
+          AlgorithmParameterSpec params) throws IOException {
     this.input = input;
     this.cipher = cipher;
     this.bufferSize = Utils.checkBufferSize(cipher, bufferSize);
-    this.key = key.clone();
-    this.initIV = iv.clone();
-    this.iv = iv.clone();
+
+    this.key = key;
+    this.params = params;
+    if (params instanceof IvParameterSpec) {
```

Comment:
We should move initIV and iv member to CTRCipherInputStream?
And if the params is not IvParameterSpec, we throw exception instead of set null values